### PR TITLE
fix(miner): fix repeatedly processing dropped txs in a new block

### PR DESCRIPTION
**Pull request by 0xmountaintop**
## 1. Purpose or design rationale of this PR

`txs.Shift()` and `txs.Pop()` will only drop the tx from the current `txs *types.TransactionsByPriceAndNonce`. However, the "dropped" tx is not really removed from the mempool, and will still be fed into `commitTransactions` when making a new block.
(You can verify this statement using https://github.com/scroll-tech/go-ethereum/pull/492. Can try and play around using https://github.com/HAOYUatHZ/ccc-tester, changing https://github.com/HAOYUatHZ/ccc-tester/blob/ccc/docker/l2geth/Dockerfile to `scrolltech/l2geth:scroll-v4.3.56-PoC`).

~~This PR removes the txs from the processing tx queue.~~ (update: This PR now removes the txs from the txpool. See discussion below.) Besides, it makes no sense to re-trace the tx, if such a single tx is already known `circuitcapacitychecker.ErrBlockRowConsumptionOverflow` or `circuitcapacitychecker.ErrUnknown`

**Note**: L1msg is not affected by this because we read and maintain the queue in db instead of reading directly from mempool.

## 2. PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [ ] build: Changes that affect the build system or external dependencies (example scopes: yarn, eslint, typescript)
- [ ] ci: Changes to our CI configuration files and scripts (example scopes: vercel, github, cypress)
- [ ] docs: Documentation-only changes
- [ ] feat: A new feature
- [ ] fix: A bug fix
- [ ] perf: A code change that improves performance
- [ ] refactor: A code change that doesn't fix a bug, or add a feature, or improves performance
- [ ] style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
- [ ] test: Adding missing tests or correcting existing tests

## 3. Deployment tag versioning

Has the version in `params/version.go` been updated?

- [ ] This PR doesn't involve a new deployment, git tag, docker image tag, and it doesn't affect traces
- [ ] Yes

## 4. Breaking change label

Does this PR have the `breaking-change` label?

- [ ] This PR is not a breaking change
- [ ] Yes

### CODE REVIEW

**Comment by 0xmountaintop** [comment]
A more performant way is to change `(pool *TxPool) removeTx` to a public function, and explictedly call it to remove the tx, instead of simply skipping it. But not sure whether that style will introduce some risks in the future.

**Comment by 0xmountaintop** [comment]
Update: I've changed my mind. We should remove the tx from the txpool. Otherwise if a user uses the nonce with `PendingNonceAt` to send a tx, the new tx will get stuck, because the previous tx is stuck in the txpool

**Comment by icemelon** [comment]
I wonder if the rpc node will remove the tx after the signer does so. Otherwise, users will get the incorrect nonce from the rpc node.

**Comment by 0xmountaintop** [comment]
> I wonder if the rpc node will remove the tx after the signer does so. Otherwise, users will get the incorrect nonce from the rpc node.

I will run a local testnet to check.

File: miner/worker.go [hunk 1]
```diff
@@ -1121,10 +1122,13 @@ loop:
 			// Circuit capacity check: unknown circuit capacity checker error for L2MessageTx, skip the account
 			log.Trace("Unknown circuit capacity checker error for L2MessageTx", "tx", tx.Hash().String())
 			log.Info("Skipping L2 message", "tx", tx.Hash().String(), "block", w.current.header.Number, "reason", "unknown row consumption error")
+			// TODO: propagate more info about the error from CCC
+			rawdb.WriteSkippedTransaction(w.eth.ChainDb(), tx, "row consumption overflow", w.current.header.Number.Uint64(), nil)
```
Comment (Thegaram) [inline miner/worker.go]:
```suggestion
			rawdb.WriteSkippedTransaction(w.eth.ChainDb(), tx, "unknown circuit capacity checker error", w.current.header.Number.Uint64(), nil)
```

**Review by Thegaram — DISMISSED** [review]

File: core/tx_pool.go [hunk 2]
```diff
@@ -1023,9 +1023,9 @@ func (pool *TxPool) Has(hash common.Hash) bool {
 	return pool.all.Get(hash) != nil
 }
 
-// removeTx removes a single transaction from the queue, moving all subsequent
+// RemoveTx removes a single transaction from the queue, moving all subsequent
 // transactions back to the future queue.
-func (pool *TxPool) removeTx(hash common.Hash, outofbound bool) {
+func (pool *TxPool) RemoveTx(hash common.Hash, outofbound bool) {
```
Comment (Thegaram) [inline core/tx_pool.go:1028]:
```suggestion
func (pool *TxPool) RemoveTx(hash common.Hash, outofbound bool) {
    pool.mu.Lock()
    defer pool.mu.Unlock()
```

safe to add locking here

Comment (Thegaram) [reply core/tx_pool.go:1028]: Or rather:

```go
func (pool *TxPool) LockAndRemoveTx(hash common.Hash) {
    pool.mu.Lock()
    defer pool.mu.Unlock()
    pool.RemoveTx(hash, true)
}
```

**Review by Thegaram — APPROVED** [review]

**Pull request merged**